File: android/src/main/java/com/github/wuxudong/rncharts/charts/VisibleEdgeAxisValueFormatter.java
package com.github.wuxudong.rncharts.charts;

import com.github.mikephil.charting.charts.BarLineChartBase;
import com.github.mikephil.charting.data.Entry;
import com.github.mikephil.charting.formatter.ValueFormatter;

public class VisibleEdgeAxisValueFormatter extends ValueFormatter {
    private final BarLineChartBase chart;
    private final ValueFormatter baseFormatter;
    private boolean enabled;

    public VisibleEdgeAxisValueFormatter(BarLineChartBase chart, ValueFormatter baseFormatter, boolean enabled) {
        this.chart = chart;
        this.baseFormatter = baseFormatter;
        this.enabled = enabled;
    }

    public ValueFormatter getBaseFormatter() {
        return baseFormatter;
    }

    public void setEnabled(boolean enabled) {
        this.enabled = enabled;
    }

    @Override
    public String getFormattedValue(float value) {
        if (!enabled) {
            return baseFormatter.getFormattedValue(value);
        }
        float lowest = chart.getLowestVisibleX();
        float highest = chart.getHighestVisibleX();

        // if the chart hasn't calculated a range yet fall back to the base
        if (highest == lowest) {
            return baseFormatter.getFormattedValue(value);
        }

        int leftIndex = (int) Math.floor(lowest);

        int rightIndex = (int) Math.ceil(highest);

        int index = Math.round(value);
        if (index == leftIndex || index == rightIndex) {
            return baseFormatter.getFormattedValue(value);
        }
        return "";
    }

    @Override
    public String getPointLabel(Entry entry) {
        return getFormattedValue(entry.getX());
    }
}

File: ios/ReactNativeCharts/formatters/VisibleEdgeAxisValueFormatter.swift
import Foundation
import DGCharts

@objc(VisibleEdgeAxisValueFormatter)
open class VisibleEdgeAxisValueFormatter: NSObject, ValueFormatter, AxisValueFormatter {
    weak var chart: BarLineChartViewBase?
    var base: AxisValueFormatter
    @objc public var enabled: Bool = true

    @objc public init(chart: BarLineChartViewBase, base: AxisValueFormatter, enabled: Bool = true) {
        self.chart = chart
        self.base = base
        self.enabled = enabled
    }

    open func stringForValue(_ value: Double, axis: AxisBase?) -> String {
        guard enabled, let chart = chart else {
            return base.stringForValue(value, axis: axis)
        }
        let handler = chart.viewPortHandler
        let leftBottom = chart.valueForTouchPoint(point: CGPoint(x: handler.contentLeft, y: handler.contentBottom), axis: .left)
        let rightTop = chart.valueForTouchPoint(point: CGPoint(x: handler.contentRight, y: handler.contentTop), axis: .left)
        let lowest = leftBottom.x
        let highest = rightTop.x

        // fall back to the base when no range is computed yet
        if lowest == highest {
            return base.stringForValue(value, axis: axis)
        }

        let leftIndex = Int(floor(lowest))
        let rightIndex = Int(ceil(highest))

        let index = Int(value.rounded())
        if index == leftIndex || index == rightIndex {
            return base.stringForValue(value, axis: axis)
        }
        return ""
    }

    open func stringForValue(_ value: Double, entry: ChartDataEntry, dataSetIndex: Int, viewPortHandler: ViewPortHandler?) -> String {
        return stringForValue(entry.x, axis: nil)
    }
}
